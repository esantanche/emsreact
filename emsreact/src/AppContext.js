// This is the application context
// We populate it in App.js and provide it to the entire application

import React from "react";

const AppContext = React.createContext(null);

export default AppContext;