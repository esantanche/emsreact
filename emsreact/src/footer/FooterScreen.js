import React, { Component } from 'react';

// Material UI components
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';

import withWidth from "@material-ui/core/withWidth/index";
import {withStyles} from "@material-ui/core/styles/index";
import FrontGridPane from '../components_library/panes/FrontGridPane';
import ColouredCard from '../components_library/cards/ColouredCard';
import SeparatorPane from '../components_library/panes/SeparatorPane';

class FooterScreen extends Component {

    render() {

        // const { classes } = this.props;

        const { width } = this.props;

        return (

            <React.Fragment>

                <FrontGridPane>

                    {/*FIXME it may be better to make a separate component for the contact me form*/}

                    <ColouredCard>

                        <CardContent>

                            <Typography variant="title" >

                                contact me form

                            </Typography>

                        </CardContent>

                    </ColouredCard>

                    <SeparatorPane />

                    <ColouredCard>

                        <CardContent>

                            <Typography variant="title" >

                                small print

                            </Typography>

                        </CardContent>

                    </ColouredCard>

                </FrontGridPane>

            </React.Fragment>

        )
    }

}

// FIXME no styles here, move them to component library
// withWidth shouldn't be used as well

export default withWidth()(FooterScreen);


