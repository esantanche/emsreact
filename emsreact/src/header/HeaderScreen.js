import React, { Component } from 'react';

// Material UI components
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';

// From components library
import MasterAppBar from "../components_library/bars/MasterAppBar";
import AlignedGridPane from "../components_library/panes/AlignedGridPane";
import HomeButtonPane from "../components_library/panes/HomeButtonPane";
import HomeButtonText from "../components_library/texts/HomeButtonText";
import HeaderMenu from "../components_library/menus/HeaderMenu";
import PickMyBrainDialog from "../components_library/dialogs/PickMyBrainDialog";
import MenuButton from "../components_library/buttons/MenuButton";
import FrontGridPane from "../components_library/panes/FrontGridPane";
import MenuButtonPane from "../components_library/panes/MenuButtonPane";

// Needed to use history and move to other url
import {withRouter} from "react-router-dom";

// Material UI function that make width available
import withWidth from "@material-ui/core/withWidth/withWidth";

/**
 * This is the header always present at the top of the page.
 *
 * There are the home button (the text "Emanuele Santanché, Web Developer") and the menu button.
 */
class HeaderScreen extends Component {

    constructor(props) {
        super(props);

        this.state = {
            // The element the menu is anchored to, typically a button that opens the menu
            // When anchorElement is null, the menu is invisible
            anchorElement: null,
            // The Pick My Brain dialog starts invisible. It will become
            // visible when the user will click on "Pick My Brain!" in the menu
            pick_my_brain_dialog_is_open: false
        };

    }

    /**
     * When the menu button, the button made of three lines, il clicked,
     * this function is called and we have to show the menu
     *
     * @param {object} event Click event for menu button
     */
    handleMenuButtonClick = event => {

        // By setting the anchor element for the menu to the menu button
        // the user has just clicked, we show the menu

        this.setState({ anchorElement: event.currentTarget });
    };

    /**
     * Called when an item of the menu is clicked.
     *
     * There may be a url to go to or we may have to show the Pick My Brain
     * dialog.
     *
     * @param {string} url_to_go_to The url we have to go to
     * @param {boolean} open_pick_my_brain_dialog True if the user clicked on
     * "Pick My Brain!" and we have to show the dialog PickMyBrainDialog
     */
    handleMenuClose = (url_to_go_to, open_pick_my_brain_dialog) => {

        this.setState({ anchorElement: null });

        if (url_to_go_to) {

            this.props.history.push(url_to_go_to);

        } else if (open_pick_my_brain_dialog) {

            this.setState({ pick_my_brain_dialog_is_open: true });

        }
    };

    /**
     * Called when the user clicks somewhere outside the menu.
     *
     * It means that we have to close the menu.
     */
    handleClickAwayFromMenu = () => {

        this.setState({ anchorElement: null });
    };

    /**
     * Called when the user clicks on the "X" to close the dialog.
     *
     * We close the dialog by setting pick_my_brain_dialog_is_open to false.
     */
    handleOnClickPickMyBrainDialog() {

        this.setState({ pick_my_brain_dialog_is_open: false });
    };

    render() {

        const { width } = this.props;

        return (

            <React.Fragment>

                <FrontGridPane>

                    <MasterAppBar>

                        <AlignedGridPane>
                            <Grid item xs={11}>
                                <HomeButtonPane>
                                    <Button href="/">

                                        {
                                            width === 'xs' ?
                                                <HomeButtonText>
                                                    Emanuele Santanché,<br /> Web Developer
                                                </HomeButtonText>
                                                :
                                                <HomeButtonText>
                                                    Emanuele Santanché, Web Developer
                                                </HomeButtonText>
                                        }

                                    </Button>
                                </HomeButtonPane>
                            </Grid>

                            <Grid item xs={1}>
                                <MenuButtonPane>
                                    <MenuButton onclick={this.handleMenuButtonClick.bind(this)} />
                                </MenuButtonPane>
                                <ClickAwayListener onClickAway={this.handleClickAwayFromMenu.bind(this)}>
                                    <HeaderMenu anchorElement={this.state.anchorElement}
                                                handleMenuClose={this.handleMenuClose.bind(this)}/>
                                </ClickAwayListener>
                            </Grid>
                        </AlignedGridPane>

                    </MasterAppBar>

                </FrontGridPane>

                <PickMyBrainDialog open={this.state.pick_my_brain_dialog_is_open}
                                   onclick={this.handleOnClickPickMyBrainDialog.bind(this)}/>

            </React.Fragment>

        )
    }

}

// withRouter is needed to be able to use the history
// like in:
//    this.props.history.push(url_to_go_to);
export default withRouter(withWidth()(HeaderScreen));

