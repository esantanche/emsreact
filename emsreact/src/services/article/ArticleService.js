// App configuration
import { APP_CONFIGURATION } from '../../appConfiguration';

// Helper function to calculate the number of the next page to fetch
// The first page is page zero
import { number_of_page_to_fetch } from '../../helpers/cache_computations';

/**
 * This service fetches articles from the backend.
 */
class ArticleService {

    constructor() {

        this.articles_cache = {};
        this.articles_count = {};

    }

    /**
     * Function getting articles.
     *
     * There are three implemented filters.
     *
     * The first one fetches a single article. It needs the node id the backend assigned to the article.
     *
     * The second fetches all articles for a given topic. Topics can be "How I work", "The tools I use", etc.
     *
     * The third fetches sticky articles only or all af them.
     *
     * It's possible to combine the topic and sticky filters. Of course the nid filter (single article) has to be used alone.
     *
     * @param {object} filters Filters to use when fetching articles
     * @param {function} callback_to_return_articles_and_more_flag Function to call to return the fetched articles
     */
    get_articles(filters, callback_to_return_articles_and_more_flag) {

        let filter_query_string = "";

        if (filters) {

            if (filters.nid)
                filter_query_string += "&nid=" + filters.nid;

            if (filters.topic)
                filter_query_string += "&field_ems_topic_target_id=" + filters.topic;

            if (filters.sticky)
                filter_query_string += "&sticky=1";

        }

        this.fetch_articles(filter_query_string, (articles_fetched_successfully) => {

            if (articles_fetched_successfully) {

                // I need to know if there are more articles to fetch so that I can return
                // this piece of information and the caller can show the LOAD MORE button

                const count_of_articles_in_cache = this.count_of_articles_cache_for_query(filter_query_string);

                const count_of_articles = this.count_of_articles_for_query(filter_query_string);

                const there_are_more_articles_to_fetch = (count_of_articles > count_of_articles_in_cache);

                const result_to_return = {

                    more: there_are_more_articles_to_fetch,

                    articles: this.articles_cache[filter_query_string]

                };

                callback_to_return_articles_and_more_flag(result_to_return);

            }

        })

    };

    /**
     * Doing the actual fetch. This function will add to the cache an additional page of results or it will do nothing
     * if all articles have already been fetched.
     * If there is nothing in cache, it will fetch page zero.
     * This function returns just true on success and false on failure. The caller will find the articles in cache.
     *
     * @param {string} filter_query_string String with query parameters that will filter the result
     * @param {function} callback Function to call to return true if everything went well, false otherwise.
     */
    fetch_articles(filter_query_string, callback) {

        let page_number;

        // There is a separated cache for each query string
        const count_of_articles_in_cache = this.count_of_articles_cache_for_query(filter_query_string);
        const count_of_articles = this.count_of_articles_for_query(filter_query_string);

        // We need to fetch another page only if we didn't yet fetch all available articles for the
        // given query.
        // If the cache is empty certainly we need to fetch the first page
        if (count_of_articles > count_of_articles_in_cache || count_of_articles_in_cache === 0) {

            if (count_of_articles === -1)
                page_number = 0;
            else
                page_number = number_of_page_to_fetch(count_of_articles_in_cache, count_of_articles, APP_CONFIGURATION.fetchPageSize);

            fetch(APP_CONFIGURATION.backendUrl + "/rest/EMS/view/articlesplayground?_format=json&langcode=en" +
                        filter_query_string + "&page=" + page_number, {
                method: 'GET',
            })
                .then((response) => {
                    if (!response.ok) { throw response }
                    return response.json()
                })
                .then((response) => {

                    this.add_results_to_cache(filter_query_string, response.results);

                    this.update_count_of_articles_for_query(filter_query_string, parseInt(response.count));

                    callback(true);
                })
                .catch((error) => {

                    callback(false);
                });


        } else {

            // Everything is already in cache. The caller has just to read it.

            callback(true);

        }

    };

    /**
     * Returning the number of articles we already have in cache for the given
     * query string.
     * We don't get the number here. We got it previously and stored it in an array.
     * Now just returning the stored value.
     *
     * @param {string} filter_query_string Query string
     * @returns {number} Number of articles in cache for the given query string
     */
    count_of_articles_cache_for_query(filter_query_string) {

        // Testing filter_query_string for empty
        // See https://repl.it/@EmanueleSantanc/Check-for-empty-parameter

        if (!filter_query_string) {

            console.error("count_of_articles_cache_for_query, filter_query_string not defined");

            return -1;
        }

        // If the cache for the given query string is not even defined,
        // we define it and set it as empty
        if (typeof this.articles_cache[filter_query_string] === "undefined") {

            this.articles_cache[filter_query_string] = [];

        }

        return this.articles_cache[filter_query_string].length;

    };

    /**
     * Returning the total number of articles that are available for the given
     * query string.
     * We don't get the number here. We got it previously and stored it in an array.
     * Now just returning the stored value.
     *
     * @param {string} filter_query_string Query string
     * @returns {number} Total number of articles that are available for the given
     * query string.
     */
    count_of_articles_for_query(filter_query_string) {

        if (!filter_query_string) {

            console.error("count_of_articles_for_query, filter_query_string not defined");

            return -1;
        }

        // If the total number of articles available for a given query has never been
        // determined, we define it and set it to -1
        if (typeof this.articles_count[filter_query_string] === "undefined") {

            this.articles_count[filter_query_string] = -1;

        }

        return this.articles_count[filter_query_string];

    };

    /**
     * We call this function to update the total number of articles available for a given query.
     * We typically do this after we make an actual query.
     *
     * @param {string} filter_query_string Query string
     * @param {number} count_of_articles Total number of articles that are available for the given
     * query string.
     */
    update_count_of_articles_for_query(filter_query_string, count_of_articles) {

        if (!filter_query_string) {

            console.error("update_count_of_articles_for_query, filter_query_string not defined");

            return;
        }

        this.articles_count[filter_query_string] = count_of_articles;

    };

    /**
     * Adding articles to the cache of articles for the given query string
     *
     * @param {string} filter_query_string Query string
     * @param {array} results Array of additional results to append to the cache
     */
    add_results_to_cache(filter_query_string, results) {

        if (typeof this.articles_cache[filter_query_string] === "undefined") {

            this.articles_cache[filter_query_string] = [];

        }

        // Append array to array
        // See https://repl.it/@EmanueleSantanc/Append-array-to-an-array

        this.articles_cache[filter_query_string].push(...results);

    }

}

export default ArticleService;