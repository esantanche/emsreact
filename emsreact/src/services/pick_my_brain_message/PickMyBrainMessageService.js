// App configuration
import { APP_CONFIGURATION } from '../../appConfiguration';

/**
 * This service gets the message the user entered in the Pick My Brain form
 * and sends it to the backend.
 *
 * Drupal will send an email to notify that there is a new message.
 */
class PickMyBrainMessageService {

    /**
     * Sending the message.
     *
     * @param {object} message_details Contains name, email and message
     * @param {function} callback_to_return_error Function to call to return the
     * error or null is everything went fine
     */
    create_pick_my_brain_message(message_details, callback_to_return_error) {

        const now = new Date();

        // The Drupal backend needs the content type of the node to create
        // It's ems_pick_my_brain_message.
        // This Drupal content type has three fields: name, email and message.
        // We have to provide a title that will show up only in Drupal
        // admin interface

        const message_node_details = {
            type: [{"target_id": "ems_pick_my_brain_message"}],
            title: [{"value": "New Message EMS " + message_details.name}],
            field_name: [{"value": message_details.name}],
            field_email: [{"value": message_details.email}],
            field_message: [{"value": message_details.message}]
        };

        fetch(APP_CONFIGURATION.backendUrl + '/node?_format=json', {
                method: 'POST',
                body: JSON.stringify(message_node_details),
                headers: {
                    'Content-Type': 'application/json',
                    'Accepts': 'application/json'
                }
            })
            .then((response) => {
                if (!response.ok) {
                    throw response
                }
                return response.json()
            })
            .then(function (response_json) {

                // No error here

                callback_to_return_error(null);
            })
            .catch(function (error) {

                console.error("PickMyBrainMessageService::create_pick_my_brain_message");
                console.error(error);

                callback_to_return_error(error);

            });

    }

}

export default PickMyBrainMessageService;
