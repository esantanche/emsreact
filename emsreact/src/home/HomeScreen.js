import React, { Component } from 'react';

// Material UI components
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Button from '@material-ui/core/Button';

// From components library
import FrontGridPane from '../components_library/panes/FrontGridPane';
import ColouredCard from '../components_library/cards/ColouredCard';
import MasterAvatar from "../components_library/avatars/MasterAvatar";

// Shows lists of articles
import ListOfArticlesScreen from '../list_of_articles/ListOfArticlesScreen';

// To provide this component with the application context
// See AppContext.js
import AppContext from '../AppContext';

/**
 * Home Screen. Header and Footer are not here.
 *
 * Here you find the presentation text ("Hello, my name is Emanuele Santanché.
 * I am a Web Developer and build websites and web applications using JavaScript,
 * React, Angular, WordPress, Drupal, HTML5, CSS3, CSSinJS.")
 * and the lists of articles separated by topic.
 *
 * The TopicService provides details about each topic.
 *
 * These details are used to show the avatar for each list of articles, the title, the subtitle
 * and the link to more articles for the same topic.
 *
 * On the home page only sticky articles are shown. Articles are flagged as sticky in the backend.
 */
class HomeScreen extends Component {

    constructor(props) {

        super(props);

        this.state = { articles: [ ], topics: [ ] };

    }

    /**
     * Fetching details about the topics to show on the home page (How I work, The tools I use, etc.).
     */
    componentDidMount() {

        const { TopicService } = this.props.services;

        this.TopicService = TopicService;

        var self = this;

        this.TopicService.get_topics(function(topics) {

            self.setState({ topics: topics });

        });

    }

    render() {

        return (

            <React.Fragment>

                <FrontGridPane>

                    <ColouredCard>

                        <CardContent>

                            <Typography variant="title" >

                                Hello, my name is Emanuele Santanché. I am a Web Developer and build websites and web applications
                                using JavaScript, React, Angular, WordPress, Drupal, HTML5, CSS3, CSSinJS.

                            </Typography>

                        </CardContent>

                    </ColouredCard>

                </FrontGridPane>

                { this.state.topics.length > 0 ?

                    this.state.topics.map((topic, index) => {

                        return (

                            <React.Fragment key={index}>

                                <FrontGridPane>

                                    <Card>
                                        <CardHeader
                                            avatar={

                                                <MasterAvatar color={topic.color_for_avatar}
                                                              letter={topic.letter_for_avatar}
                                                              label={topic.name} />


                                            }

                                            title={topic.name}
                                            subheader={topic.subheader}
                                        />

                                        <ListOfArticlesScreen topic={topic.name} sticky={true}/>

                                        <CardActions>
                                            <Button size="small"
                                                    href={"/articles/" + topic.name_for_url}>{topic.more_articles_button_label}</Button>
                                        </CardActions>

                                    </Card>

                                </FrontGridPane>

                            </React.Fragment>

                        );

                    })

                    :

                    ( <React.Fragment></React.Fragment> )

                }

            </React.Fragment>

        );
    }
}

export default props => (
    <AppContext.Consumer>
        {services => <HomeScreen {...props} services={services} />}
    </AppContext.Consumer>
);
