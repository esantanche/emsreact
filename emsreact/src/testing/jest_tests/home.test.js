// FIXME docs? anything to do?
// Testing everything in folder article
// fixme fix docs

import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import HeaderScreen from "../../header/HeaderScreen";
import {BrowserRouter} from "react-router-dom";
import AppContext from "../../AppContext";
import HomeScreen from "../../home/HomeScreen";
import PickMyBrainMessageService from "../../services/pick_my_brain_message/PickMyBrainMessageService";
import FooterScreen from "../../footer/FooterScreen";
import TopicService from "../../services/topic/TopicService";
import ArticleService from "../../services/article/ArticleService";

// FIXME this is for the home

describe('Testing HomeScreen', function() {

    const AppTopicService = new TopicService();
    const AppArticleService = new ArticleService();

    const Services = { TopicService: AppTopicService,
                       ArticleService: AppArticleService,
                       PickMyBrainMessageService: null };

    const home_screen_to_be_tested = ( <AppContext.Provider value={Services}>
                                           <HomeScreen />
                                       </AppContext.Provider> );

    it('HomeScreen renders correctly', () => {

        const tree = renderer
            .create(home_screen_to_be_tested)
            .toJSON();
        expect(tree).toMatchSnapshot();

    });

});
