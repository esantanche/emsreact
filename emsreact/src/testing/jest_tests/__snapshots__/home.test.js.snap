// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing HomeScreen HomeScreen renders correctly 1`] = `
Array [
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-23"
>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
    <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-10-66 MuiGrid-grid-lg-8-78"
    >
        <div
            className="MuiPaper-root-100 MuiPaper-elevation1-103 MuiPaper-rounded-101 MuiCard-root-99 ColouredCard-coloured_card-97"
        >
            <div
                className="MuiCardContent-root-127"
            >
                <h2
                    className="MuiTypography-root-128 MuiTypography-title-134"
                >
                    Hello, my name is Emanuele Santanch√©. I am a Web Developer and build websites and web applications using JavaScript, React, Angular, WordPress, Drupal, HTML5, CSS3, CSSinJS.
                </h2>
            </div>
        </div>
    </div>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
</div>,
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-23"
>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
    <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-10-66 MuiGrid-grid-lg-8-78"
    >
        <div
            className="MuiPaper-root-100 MuiPaper-elevation1-103 MuiPaper-rounded-101 MuiCard-root-99"
        >
            <div
                className="MuiCardHeader-root-152"
            >
                <div
                    className="MuiCardHeader-avatar-153"
                >
                    <div
                        aria-label="How I work"
                        className="MuiAvatar-root-158 MuiAvatar-colorDefault-159"
                        style={
                            Object {
                                "backgroundColor": "#f44336",
                              }
                        }
                    >
                        W
                    </div>
                </div>
                <div
                    className="MuiCardHeader-content-155"
                >
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiCardHeader-title-156"
                    >
                        How I work
                    </span>
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-157"
                    >
                        There is value in doing things properly first time
                    </span>
                </div>
            </div>
            <div
                className="MuiCardActions-root-161"
            >
                <a
                    className="MuiButtonBase-root-187 MuiButton-root-163 MuiButton-text-165 MuiButton-flat-168 MuiButton-sizeSmall-184 MuiCardActions-action-162"
                    href="/articles/how-i-work"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="button"
                    tabIndex="0"
                >
                    <span
                        className="MuiButton-label-164"
                    >
                        More about how I work
                    </span>
                    <span
                        className="MuiTouchRipple-root-190"
                    />
                </a>
            </div>
        </div>
    </div>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
</div>,
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-23"
>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
    <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-10-66 MuiGrid-grid-lg-8-78"
    >
        <div
            className="MuiPaper-root-100 MuiPaper-elevation1-103 MuiPaper-rounded-101 MuiCard-root-99"
        >
            <div
                className="MuiCardHeader-root-152"
            >
                <div
                    className="MuiCardHeader-avatar-153"
                >
                    <div
                        aria-label="The tools I use"
                        className="MuiAvatar-root-158 MuiAvatar-colorDefault-159"
                        style={
                            Object {
                                "backgroundColor": "#29b6f6",
                              }
                        }
                    >
                        T
                    </div>
                </div>
                <div
                    className="MuiCardHeader-content-155"
                >
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiCardHeader-title-156"
                    >
                        The tools I use
                    </span>
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-157"
                    >
                        Modern technology making writing code more productive, fun and easy
                    </span>
                </div>
            </div>
            <div
                className="MuiCardActions-root-161"
            >
                <a
                    className="MuiButtonBase-root-187 MuiButton-root-163 MuiButton-text-165 MuiButton-flat-168 MuiButton-sizeSmall-184 MuiCardActions-action-162"
                    href="/articles/the-tools-i-use"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="button"
                    tabIndex="0"
                >
                    <span
                        className="MuiButton-label-164"
                    >
                        More about the tools I use
                    </span>
                    <span
                        className="MuiTouchRipple-root-190"
                    />
                </a>
            </div>
        </div>
    </div>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
</div>,
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-23"
>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
    <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-10-66 MuiGrid-grid-lg-8-78"
    >
        <div
            className="MuiPaper-root-100 MuiPaper-elevation1-103 MuiPaper-rounded-101 MuiCard-root-99"
        >
            <div
                className="MuiCardHeader-root-152"
            >
                <div
                    className="MuiCardHeader-avatar-153"
                >
                    <div
                        aria-label="Success stories"
                        className="MuiAvatar-root-158 MuiAvatar-colorDefault-159"
                        style={
                            Object {
                                "backgroundColor": "#ffa726",
                              }
                        }
                    >
                        S
                    </div>
                </div>
                <div
                    className="MuiCardHeader-content-155"
                >
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiCardHeader-title-156"
                    >
                        Success stories
                    </span>
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-157"
                    >
                        From Intel Corporation to video streaming of classical music
                    </span>
                </div>
            </div>
            <div
                className="MuiCardActions-root-161"
            >
                <a
                    className="MuiButtonBase-root-187 MuiButton-root-163 MuiButton-text-165 MuiButton-flat-168 MuiButton-sizeSmall-184 MuiCardActions-action-162"
                    href="/articles/success-stories"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="button"
                    tabIndex="0"
                >
                    <span
                        className="MuiButton-label-164"
                    >
                        More success stories
                    </span>
                    <span
                        className="MuiTouchRipple-root-190"
                    />
                </a>
            </div>
        </div>
    </div>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
</div>,
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-23"
>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
    <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-10-66 MuiGrid-grid-lg-8-78"
    >
        <div
            className="MuiPaper-root-100 MuiPaper-elevation1-103 MuiPaper-rounded-101 MuiCard-root-99"
        >
            <div
                className="MuiCardHeader-root-152"
            >
                <div
                    className="MuiCardHeader-avatar-153"
                >
                    <div
                        aria-label="Testimonials"
                        className="MuiAvatar-root-158 MuiAvatar-colorDefault-159"
                        style={
                            Object {
                                "backgroundColor": "#8bc34a",
                              }
                        }
                    >
                        T
                    </div>
                </div>
                <div
                    className="MuiCardHeader-content-155"
                >
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiCardHeader-title-156"
                    >
                        Testimonials
                    </span>
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-157"
                    >
                        Happy clients and happy managers
                    </span>
                </div>
            </div>
            <div
                className="MuiCardActions-root-161"
            >
                <a
                    className="MuiButtonBase-root-187 MuiButton-root-163 MuiButton-text-165 MuiButton-flat-168 MuiButton-sizeSmall-184 MuiCardActions-action-162"
                    href="/articles/testimonials"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="button"
                    tabIndex="0"
                >
                    <span
                        className="MuiButton-label-164"
                    >
                        More testimonials
                    </span>
                    <span
                        className="MuiTouchRipple-root-190"
                    />
                </a>
            </div>
        </div>
    </div>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
</div>,
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-23"
>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
    <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-10-66 MuiGrid-grid-lg-8-78"
    >
        <div
            className="MuiPaper-root-100 MuiPaper-elevation1-103 MuiPaper-rounded-101 MuiCard-root-99"
        >
            <div
                className="MuiCardHeader-root-152"
            >
                <div
                    className="MuiCardHeader-avatar-153"
                >
                    <div
                        aria-label="Talking about my experiences"
                        className="MuiAvatar-root-158 MuiAvatar-colorDefault-159"
                        style={
                            Object {
                                "backgroundColor": "#fdd835",
                              }
                        }
                    >
                        E
                    </div>
                </div>
                <div
                    className="MuiCardHeader-content-155"
                >
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiCardHeader-title-156"
                    >
                        Talking about my experiences
                    </span>
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-157"
                    >
                        Sharing my knowledge with the world
                    </span>
                </div>
            </div>
            <div
                className="MuiCardActions-root-161"
            >
                <a
                    className="MuiButtonBase-root-187 MuiButton-root-163 MuiButton-text-165 MuiButton-flat-168 MuiButton-sizeSmall-184 MuiCardActions-action-162"
                    href="/articles/talking-about-my-experiences"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="button"
                    tabIndex="0"
                >
                    <span
                        className="MuiButton-label-164"
                    >
                        More knowledge
                    </span>
                    <span
                        className="MuiTouchRipple-root-190"
                    />
                </a>
            </div>
        </div>
    </div>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
</div>,
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-23"
>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
    <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-10-66 MuiGrid-grid-lg-8-78"
    >
        <div
            className="MuiPaper-root-100 MuiPaper-elevation1-103 MuiPaper-rounded-101 MuiCard-root-99"
        >
            <div
                className="MuiCardHeader-root-152"
            >
                <div
                    className="MuiCardHeader-avatar-153"
                >
                    <div
                        aria-label="Where? How?"
                        className="MuiAvatar-root-158 MuiAvatar-colorDefault-159"
                        style={
                            Object {
                                "backgroundColor": "#7986cb",
                              }
                        }
                    >
                        W
                    </div>
                </div>
                <div
                    className="MuiCardHeader-content-155"
                >
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiCardHeader-title-156"
                    >
                        Where? How?
                    </span>
                    <span
                        className="MuiTypography-root-128 MuiTypography-body2-136 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-157"
                    >
                        Where do I work? How do I work with you?
                    </span>
                </div>
            </div>
            <div
                className="MuiCardActions-root-161"
            >
                <a
                    className="MuiButtonBase-root-187 MuiButton-root-163 MuiButton-text-165 MuiButton-flat-168 MuiButton-sizeSmall-184 MuiCardActions-action-162"
                    href="/articles/where-how"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="button"
                    tabIndex="0"
                >
                    <span
                        className="MuiButton-label-164"
                    >
                        More about where and how I work with you
                    </span>
                    <span
                        className="MuiTouchRipple-root-190"
                    />
                </a>
            </div>
        </div>
    </div>
    <div
        className="MuiGrid-item-2 MuiGrid-grid-md-1-57 MuiGrid-grid-lg-2-72"
    />
</div>,
]
`;
